### ISSUE TITLE
Getting unexpected login dialog box

### ISSUE THREAD

**Issue by pashvin** [AzureAD/microsoft-authentication-library-for-js#4056]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.13.1

### Wrapper Library

Not Applicable

### Wrapper Library Version

NA

### Description

App is using latest MSAL 2.0 library and passing token to Graph library using acquireTokenSilent. Randomly acquireTokenSilent is returning "InteractionRequiredAuthError" which ends up calling acquireTokenPopup and user sees login dialog box.

I understand that if refresh token is expired (which should happen only after 90 days), then silent token method may fail but some user are getting within  4/5 days after initial login and getting fresh token.

I feel SDK is sometime randomly sending "InteractionRequiredAuthError" error even though refresh token is valid.

We also verified that 3rd party cookies are not blocked (if it is blocked I actually get "InteractionRequiredAuthError" error within 24 hours).

### MSAL Configuration

```javascript
this.msalauthService = new PublicClientApplication({
      auth: {
        clientId: this.getClientId(),
        redirectUri: window.location.origin + '/office365/login/redirect',
      },
      cache: {
        cacheLocation: 'localStorage',
        storeAuthStateInCookie: false, // set to true for IE 11
      },
      system: {
        iframeHashTimeout: 10000,
      },
    });
```

### Relevant Code Snippets

```javascript
private initGraphClient() {
    this.graphClient = MSGraphClient.init({
      authProvider: async (done) => {
        // Get the token from the auth service
        let token = await this.getAccessToken().catch((reason) => {
          done(reason, null);
        });

        if (token) {
          done(null, token);
        } else {
          done('Could not get an access token', null);
        }
      },
    });
  }

  private async getAccessToken(): Promise<string> {
    let InteractionRequiredAuth = false;
    let result = await this.msalauthService.acquireTokenSilent({ scopes: this.tokenScope }).catch((reason) => {
      if (reason + ''.includes('InteractionRequiredAuthError')) {
        Logging.info('Microsoft Enterprise Account: Interaction required ');
        InteractionRequiredAuth = true;
      } else {
        Logging.error('Microsoft Enterprise Account: ms token aquire first trial failed:' + reason);
      }
    });

    if (InteractionRequiredAuth) {
      // use aquiretoken popup instead of login popup to avoid
      // issue of poping login window every day.
      result = await this.msalauthService.acquireTokenPopup({ scopes: this.tokenScope }).catch((reason) => {
        Logging.error('Microsoft Enterprise Account: ms token aquire second trial failed:' + reason);
      });
    }

    if (result) {
      return result.accessToken;
    }
    return '';
  }
```

### Reproduction Steps

1.Login into O365
2.keep using app every day to make sure token are renewed every day
3.wait for "InteractionRequiredAuthError" within few days (it is random but happen within few days which is less than 90days) 

### Expected Behavior

If user is keep using client, then error should not happen ever because it should renew refresh token every day but in worst case error should not happen before 90days from the first login.

### Identity Provider

ADFS

### Browsers Affected (Select all that apply)

Chrome

### Source

External (Customer)

**Comment by tnorling** [comment]
@pashvin Refresh tokens issued to "SPA" apps have a 24 hour, non adjustable lifetime as a security consideration given that the tokens are stored in the browser. This is documented [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/token-lifetimes.md#refresh-tokens). When the RT is expired MSAL.js will attempt to acquire a new set of access and refresh tokens using your session cookies on the service, which explains why this happens after a few days when 3rd party cookies are not blocked, but after only 24 hours when they are blocked.

**Comment by pashvin** [comment]
@tnorling  thanks for the quick response

How this library is getting 24 hours token because as per MS document this RT timeout is fixed 90 days or can be adjusted by only AD admin for their tenant (and we don't have any custom setting so it should be 90 days for us)

Ref - https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-configurable-token-lifetimes

Also if library is defaulting to 24 hours , then is there any way for MSAL library consumer to increase refresh token timeout using any SDK settings because 24 hours is lot less.

**Comment by tnorling** [comment]
@pashvin This is dictated by the service for all reply urls of type "SPA" (which is required to use MSAL.js v2 or any other client-side auth library) and is not configurable. See the note in the [refresh token document](https://docs.microsoft.com/en-us/azure/active-directory/develop/refresh-tokens#token-timeouts). If you need an RT with a longer lifetime you'll need to use a server side library (msal-node, msal.net, etc) and do your authentication server-side. Client-side auth in the browser only supports 24 hour RTs.

**Comment by pashvin** [comment]
@tnorling thanks for confirmation that RT is only for 24 hours. I also got some more detail from one of the user. The user is putting his laptop in sleep over night while application running and then weak-up in morning which is less than 24 hours but I see in that case MSAL is sending "InteractionRequiredAuthError" while renewing token in morning. Is it possible is we try to call acquireTokenSilent while network is not ready (it takes some times on some laptop to connect with network after weakup) may fail renew token and then ask for interactive token renewal?

Anyway I am going to add verbose logs for that user and can find out more detail.

**Comment by tnorling** [comment]
@pashvin No, `InteractionRequiredAuthError` is thrown as a result of very specific error codes returned by the server (token expired, consent required, MFA required, login required, etc.) If the network is unavailable you should get a [no_network_connectivity error](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/a653fab258bb9abcd5c34e32aaf13a6999258287/lib/msal-browser/src/error/BrowserAuthError.ts#L132) or [get_request_failed/post_request_failed error](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/src/error/BrowserAuthError.ts#L140).

To further clarify the RT is valid for 24 hours from the time it was obtained and **is not renewable**. If you make an `acquireTokenSilent` call after the RT has expired MSAL.js will attempt to acquire a new 24 hr RT silently (if 3rd party cookies are blocked this will fail). You can also obtain a new 24 hr RT when you make an interactive request. To further illustrate this: if your user logs in at 9am and only makes `acquireTokenSilent` requests all day the RT will expire at 9am the following day. At this point if 3rd party cookies are blocked or the server session has expired `acquireTokenSilent` will throw an `InteractionRequiredAuthError` which you should handle by invoking interaction.

If you are certain the original RT was obtained less than 24 hours ago and the server is still expecting interaction it may be expecting a different type of interaction, such as consent. You can email me a fiddler trace if you're unsure what kind of interaction is expected.

**Comment by pashvin** [comment]
@tnorling Yes, I understand that RT is valid from the time it is obtained but here is my understanding First time login MSAL obtain access token (1 hour validity) and refresh token (24 hour validity) and app continue to use access token till hour to make MS graph calls. After 1 hour MS Graph request will find access token is not valid and that that point we request acquireTokenSilent and that gets new access token and refresh token (which should be valid for next 24 hour) so if user is just putting machine in sleep overnight, then RT should be valid. If user machine down little over 24 hour, then RT will be invalid. 

So  does acquireTokenSilent renew (or get new one) refresh token or not?  if not then is there any way to force it?

**Comment by pashvin** [comment]
I am assuming acquireTokenSilent is getting new RT because if that is not the case , then most of user must be reporting for interactive prompt after every weekend because many of them not running app in weekend and that is more than 24 hours.

**Comment by tnorling** [comment]
@pashvin No, again the RT is **not renewable**, `acquireTokenSilent` does not obtain a new RT on every call. There are 3 ways to get a new RT:

1. Call an interactive acquireToken API
2. Call `ssoSilent` API
3. Call `acquireTokenSilent` API **after** RT has expired

Methods 2 and 3 above are essentially the same as `acquireTokenSilent` calls `ssoSilent` under the hood when it needs a new RT. However, these methods will not always succeed and you should fallback to interaction when they throw `InteractionRequiredAuthError`. A few reasons why they might fail are: 

1. 3rd party cookies are blocked in the browser
2. The user does not have an existing, valid session cookie from the server
3. The user has not previously consented to the scopes requested
4. Multi factor authentication is required

Our recommendation is to call `acquireTokenSilent` each time you need a token and fallback to interaction when you receive `InteractionRequiredAuthError`. Unfortunately, if your goal is to avoid the `InteractionRequiredAuthError` in the first place, that won't be possible. Users can reduce the frequency with which they are prompted for interaction beyond 24 hours by ensuring 3P cookies are not blocked in their browser but they will still be asked to sign in again from time to time when their server session expires.

**Comment by pashvin** [comment]
@tnorling most of our user has MFA enabled and they are in AD network so do you think when "acquireTokenSilent" fails with Interaction Required error (assuming user consent is done), then instead of calling acquireTokenPopup directly , if I call ssoSilent first, then it may reduce frequency of popup?

Also now I know that this popup may come and in some cases I can't avoid it, then is there any way to find out that popup already open and waiting for user input because I don't have that check and sometime more than one popup are opening (I have code which just calls acquireTokenPopup on InteractionRequiredAuth error

**Comment by tnorling** [comment]
@pashvin 

> so do you think when "acquireTokenSilent" fails with Interaction Required error (assuming user consent is done), then instead of calling acquireTokenPopup directly , if I call ssoSilent first, then it may reduce frequency of popup?

No, if you receive `InteractionRequiredAuthError` you must invoke interaction to resolve. `ssoSilent` will also fail.

> is there any way to find out that popup already open and waiting for user input because I don't have that check and sometime more than one popup are opening

MSAL.js should throw an `interaction_in_progress` error if you try to invoke `acquireTokenPopup` while another is already in progress. When this happens do you perhaps have multiple tabs/windows open that may each be invoking their own popup? Can you please share a .gif showing multiple popups opening?

The API returns a promise, if you need to block other code from running until it's complete you can await the promise.

**Comment by pashvin** [comment]
acquireTokenPopup does throw interaction_in_progress but it also shows popup so it is too late because I see those error in log but also see multiple popup. Do you think it should just throw error and not show popup and it is MSAL bug?

I am thinking to introduce some flag on my side to avoid duplicate unless there is some flag already which I can use it.

**Comment by pashvin** [comment]
Few more update from my app users. Some users are saying they get popup (and based on our discussion that is due to access and RT both expired) and some user just tried to just close popup and they are able to get data (my app gets user meetings using graph API which uses token from MSAL). I don't understand how the token is refreshed without user respond to interactive popup?

**Comment by tnorling** [comment]
> acquireTokenPopup does throw interaction_in_progress but it also shows popup so it is too late because I see those error in log but also see multiple popup. Do you think it should just throw error and not show popup and it is MSAL bug?

It should just throw the error and not open a 2nd popup. Can you please provide a video/.gif of this behavior and/or a sample app that reproduces this?

> Some users are saying they get popup (and based on our discussion that is due to access and RT both expired) and some user just tried to just close popup and they are able to get data (my app gets user meetings using graph API which uses token from MSAL). I don't understand how the token is refreshed without user respond to interactive popup?

If a user closes the popup the API will throw a `user_cancelled` error. I can't speculate as to what's going on here, I would need a sample app that reproduces this behavior to debug further.

**Comment by pashvin** [comment]
I don't have small sample app yet but I have some internal product screenshot which shows two dialog box (and logs shows two fast calls and then exception and later each call has exception). Most of user reported two and not more than that. Here is log statement which I quickly placed for one user to track down. I don't know when and how the duplicate popup check is done but looks like it may be skipping some early calls if those happen too fast.

"2021-09-13T08:23:45.858Z INFO App Microsoft Enterprise Account: Interaction required  ",
"2021-09-13T09:25:13.073Z INFO App Microsoft Enterprise Account: Interaction required  ",
"2021-09-13T09:25:13.077Z ERROR App Microsoft Enterprise Account: ms token aquire second trial failed:BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: microsoft-authentication-library-for-js/errors.md at dev · AzureAD/microsoft-authentication-library-for-js . ",
"2021-09-13T11:26:32.062Z INFO App Microsoft Enterprise Account: Interaction required  ",
"2021-09-13T11:26:32.063Z ERROR App Microsoft Enterprise Account: ms token aquire second trial failed:BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: microsoft-authentication-library-for-js/errors.md at dev · AzureAD/microsoft-authentication-library-for-js . ",
"2021-09-13T11:26:35.980Z INFO App Microsoft Enterprise Account: Interaction required  ",
"2021-09-13T11:26:35.980Z ERROR App Microsoft Enterprise Account: ms token aquire second trial failed:BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: microsoft-authentication-library-for-js/errors.md at dev · AzureAD/microsoft-authentication-library-for-js . ",
"2021-09-13T12:21:19.040Z INFO App Microsoft Enterprise Account: Interaction required  ",
"2021-09-13T12:21:19.128Z ERROR App Microsoft Enterprise Account: ms token aquire second trial failed:BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: microsoft-authentication-library-for-js/errors.md at dev · AzureAD/microsoft-authentication-library-for-js . ",
"2021-09-13T12:47:32.484Z INFO App Microsoft Enterprise Account: Interaction required  ",
"2021-09-13T12:47:32.539Z ERROR App Microsoft Enterprise Account: ms token aquire second trial failed:BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: microsoft-authentication-library-for-js/errors.md at dev · AzureAD/microsoft-authentication-library-for-js . ",
"2021-09-13T12:47:36.040Z INFO App Microsoft Enterprise Account: Interaction required  ",
"2021-09-13T12:47:36.090Z ERROR App Microsoft Enterprise Account: ms token aquire second trial failed:BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: microsoft-authentication-library-for-js/errors.md at dev · AzureAD/microsoft-authentication-library-for-js . ",
"2021-09-13T13:46:02.964Z INFO App Microsoft Enterprise Account: Interaction required  ",
"2021-09-13T13:46:02.965Z ERROR App Microsoft Enterprise Account: ms token aquire second trial failed:BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: microsoft-authentication-library-for-js/errors.md at dev · AzureAD/microsoft-authentication-library-for-js . ",

**Comment by pashvin** [comment]
When user cancel dialog box, then I see "user_cancelled" error in the log but also see calls to graph API is fine.

By the way is there any impact on "Keep me signed" checked by user for 24 hour RT expiration as mentioned in following web page.

Ref: https://www.zubairalexander.com/blog/azure-active-directorys-configurable-token-lifetimes/

**Comment by tnorling** [comment]
One of the first things `acquireTokenPopup` does is checks sessionStorage for an `interaction_in_progress` value. If it's there it throws and if it's not it creates it. If you make 2 acquireTokenPopup calls at virtually the same time it is possible that they both make it through that check and open their own popups. Are you sure the `interaction_in_progress` error you are seeing isn't being thrown by a 3rd or 4th, etc. invocation of `acquireTokenPopup`? In any event you should debug your app to find out why `acquireTokenPopup` is being invoked multiple times back to back and suppress that behavior.

> When user cancel dialog box, then I see "user_cancelled" error in the log but also see calls to graph API is fine.

At what point is the popup closed? Is it expecting credentials when the user closes it? Or has the user already entered their credentials? Like I said, it's hard to speculate without a reproduction of this behavior, but the most reasonable I can think of is that the user successfully renewed their server session before closing the popup and subsequent `acquireTokenSilent` calls in the background start succeeding using the newly renewed session cookie.

> By the way is there any impact on "Keep me signed" checked by user for 24 hour RT expiration as mentioned in following web page.

KMSI only affects the session cookie, not the RT. You may see improvements in the frequency with which you are prompted via popup if you check the KMSI box as it will extend the server session. However, this will have no affect in browsers that block 3P cookies as this session cookie is used after the 24 hour RT expiration and will be unavailable in those browsers.

**Comment by pashvin** [comment]
@tnorling By the way thanks for all your valuable time and advice. Appreciate your prompt response.

For multiple popup , yes may be my app is making quick multiple call graph call on startup which may be getting integration required error and creating race condition and I will fix that in my app.

For canceling popup, user are not entering username or password, they are just closing it. I am adding verbose level log so will see if that gives any detail.

For 24 hour RT expiration, if I obtain refresh token my self (by renewing every hour), then is there any way I can update and hand over new RT to MSAL lib?

I have one old app (written before MSAL lib) where I was renewing refresh token with access token so may be I can use that trick to renew refresh token and save in the storage where MSAL can use it.

oAuth_refreshAccessToken: function() {
+        /* jshint ignore:start */
+        var dfd = can.Deferred();
+        var url = 'https://login.microsoftonline.com/common/oauth2/v2.0/token';
+        
+        fetch(url,
+        {
+            crossDomain: true,
+            headers: {
+            'Accept': 'application/json',
+            'Content-Type': 'application/x-www-form-urlencoded'
+            },
+            method: 'POST',
+            body: 'client_id=test-id&scope=https%3A%2F%2Foutlook.office.com%2FEWS.AccessAsUser.All&refresh_token='+localStorage.getItem('microsoft_refresh_token')+'&grant_type=refresh_token'   
+        })
+        .then(response => response.json())
+        .then(responseJson => {
+            if(responseJson.access_token !== undefined && responseJson.refresh_token !== undefined) {
+                console.log('Microsoft OAuth: Renewing Access and Refresh token');
+                localStorage.setItem('microsoft_access_token',responseJson.access_token );

**Comment by tnorling** [comment]
> For 24 hour RT expiration, if I obtain refresh token my self (by renewing every hour), then is there any way I can update and hand over new RT to MSAL lib?

There is not.

> I have one old app (written before MSAL lib) where I was renewing refresh token with access token so may be I can use that trick to renew refresh token and save in the storage where MSAL can use it.

What you have here is very similar to what MSAL does to renew the access token. I would like to reiterate that the refresh token behavior is **not** an MSAL.js limitation but rather a result of very deliberate protocol designs implemented on the server to mitigate several attack vectors. You can use any Oauth compliant library you wish or even write your own, but as long as you are targeting AAD with a replyUrl of type "SPA" you will receive a 24 hr fixed lifetime RT. The only way to get a fresh RT is to make a new call to the /authorize endpoint which requires either cookies or interaction. The /token endpoint will issue a "new" RT but its lifetime will be equal to the remaining lifetime of the original RT. I would caution you not to try to work around this as these decisions were made for really important security related reasons. If you did manage to get a new 24 hr RT from the /token endpoint with replyUrl of type "SPA" please report it to [MSRC](https://aka.ms/report-security-issue) as that is something we would certainly want to fix.

**Comment by pashvin** [comment]
Ok I understand 24 hour rule and it is MS enforced but user are reporting before 24 hours. Now myself can't reproduce so it is working for me beyond 24 hours so I have few choices
1) spend time to create sample app and reproduce and as I said it never happens on my machine so will not have scucess
2) in old app without using MSAL don't have issue so find out how that works but not MSAL lib.
3) generate token on server side which is 90days and pass to MSAL

**Comment by tnorling** [comment]
@pashvin 

> Ok I understand 24 hour rule and it is MS enforced but user are reporting before 24 hours.

If it's happening before 24 hours it's very likely due to something other than RT expiration. Maybe the RT was removed from browser storage or MFA is required. Best way to confirm is by collecting network traces or providing me with a correlationId and timestamp when this happens.

> generate token on server side which is 90days and pass to MSAL

If your app has a server side component that can acquire tokens you should be acquiring tokens server side using a server side library such as msal-node. Server side flows are not subject to the same restrictions that browser based flows are. Please do not acquire tokens server side and store in the browser, especially long lived refresh tokens.

**Comment by pashvin** [comment]
We have all user enabled with MFA but only few has this issue. I asked one user who can reproduce always to capture new logs.

**Comment by pashvin** [comment]
While we are waiting for user logs, I have one more question. Is there any way to get popup window object handle? Sometime this popup comes at startup and it goes behind the main app (because user may accidentally click on main window) so popup is waiting for user input and user has main window and it is waiting for data. If we get handle on popup , then I can register on unfocus event and put back in focus.

**Comment by tnorling** [comment]
You can use the event API and listen to the `POPUP_OPENED` event. The payload is the popup window object. You can read more about the event API usage [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/events.md)

**Comment by pashvin** [comment]
Yes, I have hook for those event but those event does not provide popup window object so I can't hookup to find out popup is going in/out focus or can't put back in focus. I am looking a way to control popup and put in user visibility or make it always on top so user sees it.

**Comment by tnorling** [comment]
Can you please share the code you are using to listen to the event? Like I said the payload for the `POPUP_OPENED` event is the popup window object. [See here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/a653fab258bb9abcd5c34e32aaf13a6999258287/lib/msal-browser/src/interaction_client/PopupClient.ts#L110)

**Comment by pashvin** [comment]
It was my mistake. I was adding following code in LOGIN START event. I see popup window object in POPUP_OPENED event but still can't force to focus because it is a cross origin popup so I don't get blur event in my handler so I don't think I can put in focus.

this.msalauthService.addEventCallback((event: EventMessage) => {
if (event.eventType === EventType.POPUP_OPENED) {
          const popup: Window = event.payload['popupWindow'];
          if(popup) {
            popup.onblur = function(event) { 
              popup.focus();
            }
          }
        }
}

**Comment by pashvin** [comment]
@tnorling I got one network capture from user but during overnight sleep test there was internet outage so if this has not good info, then I am going to get one more clean test tomorrow.
[test-1.har.zip](https://github.com/AzureAD/microsoft-authentication-library-for-js/files/7173743/test-1.har.zip)

**Comment by tnorling** [comment]
Did this fail due to `InteractionRequired`? I'm only seeing a gateway timeout (likely due to the internet outage) which does not result in `InteractionRequired`. 

Side note: please email me (email on my profile) the next trace instead of posting publicly due to the sensitive nature of what is inside.

**Comment by pashvin** [comment]
So user put machine in the sleep overnight and in morning he saw MS login window but that was blank and then internet connection was back and popup window got login ui so it got the popup (which normally happen due to interaction required) but anyway I don't want to go in wrong direction because of internet outage so if something does not make sense, then don't spend time and will get next set of logs tomorrow.

**Comment by pashvin** [comment]
@tnorling sent some new logs to your email address.

**Comment by pashvin** [comment]
Still under monitor and waiting for new logs

**Comment by pashvin** [comment]
Closing issue as after upgrading to 2.17 issue is not noticed
